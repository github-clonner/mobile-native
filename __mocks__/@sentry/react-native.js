export const captureMessage = jest.fn();
export default {};
