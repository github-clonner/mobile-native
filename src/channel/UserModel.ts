import { observable, action, runInAction } from 'mobx';

import { MINDS_CDN_URI, GOOGLE_PLAY_STORE } from '../config/Config';
import api from '../common/services/api.service';
import BaseModel from '../common/BaseModel';
import ChannelService from './ChannelService';
import sessionService from '../common/services/session.service';
import apiService from '../common/services/api.service';
import logService from '../common/services/log.service';

//@ts-nocheck
export const USER_MODE_OPEN = 0;
export const USER_MODE_MODERATED = 1;
export const USER_MODE_CLOSED = 2;

/**
 * User model
 */
export default class UserModel extends BaseModel {
  merchant;
  /**
   * Eth wallet
   */
  eth_wallet: string = '';
  wire_rewards;
  sums;
  btc_address?: string;
  icontime!: string;
  username!: string;
  name!: string;
  admin = false;
  plus: boolean = false;
  verified: boolean = false;
  founder: boolean = false;
  rewards: boolean = false;
  last_accepted_tos: number = 0;
  carousels?: Array<any>;

  /**
   * @var {boolean}
   */
  @observable blocked;

  /**
   * @var {number}
   */
  @observable subscribers_count;

  /**
   * @var {number}
   */
  @observable impressions;

  /**
   * @var {boolean}
   */
  @observable subscribed;

  /**
   * @var {boolean}
   */
  @observable mature_visibility = false;

  /**
   * @var {boolean}
   */
  @observable pending_subscribe = false;

  /**
   * @var {numeric}
   */
  @observable mode = 0;

  /**
   * @var {boolean}
   */
  @observable email_confirmed = false;

  /**
   * Confirm email
   * @param {Object} params
   */
  confirmEmail = async (params) => {
    // call any api endpoint with the param
    try {
      await apiService.get('api/v2/entities/', { urn: this.urn, ...params });
      this.setEmailConfirmed(true);
      return true;
    } catch (error) {
      return false;
    }
  };

  /**
   * Get the user icon time
   */
  getOwnerIcontime(): string {
    if (sessionService.getUser().guid === this.guid) {
      return sessionService.getUser().icontime;
    } else {
      return this.icontime;
    }
  }

  @action
  toggleMatureVisibility() {
    if (GOOGLE_PLAY_STORE) return;
    this.mature_visibility = !this.mature_visibility;
  }

  @action
  async toggleSubscription() {
    const value = !this.subscribed;
    this.subscribed = value;
    try {
      const metadata = this.getClientMetadata();
      await ChannelService.toggleSubscription(this.guid, value, metadata);
      UserModel.events.emit('toggleSubscription', this);
    } catch (err) {
      runInAction(() => {
        this.subscribed = !value;
      });
      throw err;
    }
  }

  @action
  async toggleBlock(value: boolean | null = null) {
    value = value === null ? !this.blocked : value;

    try {
      await ChannelService.toggleBlock(this.guid, value);
      this.blocked = value;
    } catch (err) {
      this.blocked = !value;
      logService.exception('[ChannelStore] toggleBlock', err);
    }
  }

  @action
  setMode(value) {
    this.mode = value;
  }

  @action
  setEmailConfirmed(value) {
    this.email_confirmed = value;
  }

  /**
   * Is admin
   */
  isAdmin() {
    return this.admin;
  }

  /**
   * current user is owner of the channel
   */
  isOwner = () => {
    return sessionService.getUser().guid === this.guid;
  };

  /**
   * Get banner source
   */
  getBannerSource() {
    if (this.carousels) {
      return {
        uri: this.carousels[0].src,
      };
    }
    return {
      uri: `${MINDS_CDN_URI}fs/v1/banners/${this.guid}/fat/${this.icontime}`,
      headers: api.buildHeaders(),
    };
  }

  /**
   * Get avatar source
   * @param {string} size
   */
  getAvatarSource(size = 'medium') {
    return {
      uri: `${MINDS_CDN_URI}icon/${
        this.guid
      }/${size}/${this.getOwnerIcontime()}`,
      headers: api.buildHeaders(),
    };
  }

  /**
   * Has banner
   * @returns {boolean}
   */
  hasBanner() {
    return !!this.carousels;
  }

  /**
   * Is closed
   * @returns {boolean}
   */
  isClosed() {
    return this.mode === USER_MODE_CLOSED;
  }

  /**
   * Is open
   * @returns {boolean}
   */
  isOpen() {
    return this.mode === USER_MODE_OPEN;
  }

  /**
   * Is moderated
   * @returns {boolean}
   */
  isModerated() {
    return this.mode === USER_MODE_MODERATED;
  }

  /**
   * Is subscribed
   * @returns {boolean}
   */
  isSubscribed() {
    return !!this.subscribed;
  }

  /**
   * Request subscribe
   */
  async subscribeRequest() {
    if (this.pending_subscribe || this.mode !== USER_MODE_CLOSED) {
      return;
    }
    try {
      this.pending_subscribe = true;
      await apiService.put(`api/v2/subscriptions/outgoing/${this.guid}`);
    } catch (err) {
      this.pending_subscribe = false;
      logService.exception(err);
    }
  }

  /**
   * Cancel subscribe request
   */
  async cancelSubscribeRequest() {
    if (!this.pending_subscribe || this.mode !== USER_MODE_CLOSED) {
      return;
    }
    try {
      this.pending_subscribe = false;
      await apiService.delete(`api/v2/subscriptions/outgoing/${this.guid}`);
    } catch (err) {
      this.pending_subscribe = true;
      logService.exception(err);
    }
  }
}
