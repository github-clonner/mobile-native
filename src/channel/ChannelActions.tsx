//@ts-nocheck
import React, { Component } from 'react';

import { View, ActivityIndicator } from 'react-native';

import { observer } from 'mobx-react';

import Icon from 'react-native-vector-icons/Ionicons';
import i18n from '../common/services/i18n.service';
import ActionSheet from 'react-native-actionsheet';
import featuresService from '../common/services/features.service';
import sessionService from '../common/services/session.service';
import Button from '../common/components/Button';
import withPreventDoubleTap from '../common/components/PreventDoubleTap';
import { CommonStyle as CS } from '../styles/Common';
import {
  FLAG_SUBSCRIBE,
  FLAG_MESSAGE,
  FLAG_EDIT_CHANNEL,
  FLAG_WIRE,
} from '../common/Permissions';
import ChannelModeSelector from './ChannelModeSelector';
import ThemedStyles from '../styles/ThemedStyles';

const ButtonCustom = withPreventDoubleTap(Button);

/**
 * Channel Actions
 */
@observer
class ChannelActions extends Component {
  state = {};

  componentDidMount() {
    this.getScheduledCount();
  }

  showActionSheet = () => {
    this.ActionSheet.show();
  };

  handleSelection = (index) => {
    this.executeAction(index);
  };

  getOptions() {
    let options = [i18n.t('cancel')];

    if (this.props.store.channel.isSubscribed()) {
      options.push(i18n.t('channel.unsubscribe'));
    }

    if (!this.props.store.channel.blocked) {
      options.push(i18n.t('channel.block'));
    } else {
      options.push(i18n.t('channel.unblock'));
    }

    options.push(i18n.t('channel.report'));

    return options;
  }

  executeAction(option) {
    let options = this.getOptions();
    let selected = options[option];
    switch (selected) {
      case i18n.t('channel.unsubscribe'):
        this.props.store.channel.toggleSubscription();
        break;
      case i18n.t('channel.block'):
        this.props.store.channel.toggleBlock();
        break;
      case i18n.t('channel.unblock'):
        this.props.store.channel.toggleBlock();
        break;
      case i18n.t('channel.report'):
        this.props.navigation.push('Report', {
          entity: this.props.store.channel,
        });
        break;
    }
  }

  toggleSubscription = () => {
    this.props.store.channel.toggleSubscription();
  };

  /**
   * Navigate To conversation
   */
  navToConversation = () => {
    if (this.props.navigation) {
      this.props.navigation.push('Conversation', {
        conversation: {
          guid: this.props.store.channel.guid + ':' + sessionService.guid,
        },
      });
    }
  };

  openWire = () => {
    this.props.navigation.navigate('WireFab', {
      owner: this.props.store.channel,
    });
  };

  onEditAction = () => {
    this.props.onEditAction();
  };

  onViewScheduledAction = async () => {
    this.props.onViewScheduledAction();
  };

  getScheduledCount = async () => {
    if (featuresService.has('post-scheduler')) {
      await this.props.store.feedStore.getScheduledCount();
    }
  };

  shouldRenderScheduledButton = () => {
    return featuresService.has('post-scheduler') && !this.state.edit;
  };

  setSheetRef = (o) => (this.ActionSheet = o);

  /**
   * Render Header
   */
  render() {
    const channel = this.props.store.channel;
    const isOwner = channel.isOwner();
    const showWire =
      !channel.blocked &&
      !isOwner &&
      featuresService.has('crypto') &&
      channel.can(FLAG_WIRE);
    const showScheduled =
      featuresService.has('post-scheduler') && !this.state.edit && isOwner;
    const showSubscribe =
      !isOwner && !channel.isSubscribed() && channel.can(FLAG_SUBSCRIBE);
    const showMessage =
      !isOwner && channel.isSubscribed() && channel.can(FLAG_MESSAGE);
    const showEdit = isOwner && channel.can(FLAG_EDIT_CHANNEL);
    const showMode =
      isOwner && featuresService.has('permissions') && this.props.editing;

    if (this.props.store.isUploading) {
      return <ActivityIndicator size="small" />;
    }

    return (
      <View style={[CS.rowJustifyEnd, CS.marginTop2x]}>
        {showScheduled && (
          <ButtonCustom
            onPress={this.onViewScheduledAction}
            accessibilityLabel={i18n.t('channel.viewScheduled')}
            text={`${i18n.t('channel.viewScheduled')}: ${
              this.props.store.feedStore.feedStore.scheduledCount
            }`}
            loading={this.state.saving}
            inverted={
              this.props.store.feedStore.endpoint ==
              this.props.store.feedStore.scheduledEndpoint
                ? true
                : undefined
            }
          />
        )}
        {showSubscribe && (
          <ButtonCustom
            onPress={this.toggleSubscription}
            containerStyle={[CS.rowJustifyCenter, CS.marginLeft]}
            accessibilityLabel={i18n.t('channel.subscribeMessage')}
            text={i18n.t('channel.subscribe')}
            testID="SubscribeButton"
          />
        )}
        {showMessage && (
          <ButtonCustom
            onPress={this.navToConversation}
            containerStyle={[CS.rowJustifyCenter, CS.marginLeft]}
            accessibilityLabel={i18n.t('channel.sendMessage')}
            text={i18n.t('channel.message')}
            testID="SendMessageButton"
          />
        )}
        {showEdit && (
          <ButtonCustom
            onPress={this.onEditAction}
            containerStyle={[CS.rowJustifyCenter, CS.marginLeft]}
            accessibilityLabel={
              this.props.editing
                ? i18n.t('channel.saveChanges')
                : i18n.t('channel.editChannel')
            }
            text={this.props.editing ? i18n.t('save') : i18n.t('edit')}
            loading={this.props.saving}
            testID="EditButton"
          />
        )}
        {showMode && <ChannelModeSelector channel={channel} />}
        {showWire && (
          <ButtonCustom
            onPress={this.openWire}
            accessibilityLabel="Wire Button"
            containerStyle={[CS.rowJustifyCenter, CS.marginLeft]}
            textStyle={[CS.marginLeft, CS.marginRight]}
            icon="ios-flash"
            text="Wire"
            testID="WireButton">
            <Icon
              name="ios-flash"
              size={16}
              style={[CS.marginLeft, ThemedStyles.style.colorPrimaryText]}
            />
          </ButtonCustom>
        )}
        {!isOwner && (
          <ButtonCustom
            onPress={this.showActionSheet}
            accessibilityLabel={i18n.t('more')}
            containerStyle={[CS.rowJustifyCenter, CS.marginLeft]}
            textStyle={[CS.marginLeft, CS.marginRight]}
            icon="ios-flash"
            text={i18n.t('more')}
            testID="MoreButton"
          />
        )}
        <ActionSheet
          ref={this.setSheetRef}
          title={i18n.t('actions')}
          options={this.getOptions()}
          onPress={this.handleSelection}
          cancelButtonIndex={0}
        />
      </View>
    );
  }
}

export default ChannelActions;
