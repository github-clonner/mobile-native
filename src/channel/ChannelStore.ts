//@ts-nocheck
import { observable, action } from 'mobx';
import { cloneDeep } from 'lodash';

import channelService from './ChannelService';
import wireService from '../wire/WireService';
import ChannelFeedStore from './ChannelFeedStore';
import UserModel from './UserModel';
import logService from '../common/services/log.service';
import channelsService from '../common/services/channels.service';

/**
 * Channel Store
 */
export default class ChannelStore {
  guid = null;
  feedStore = null;

  // only observable by reference because UserModel already have the needed observables
  @observable.ref channel = {};

  @observable rewards = {};
  @observable active = false;
  @observable loaded = false;
  @observable isUploading = false;
  @observable avatarProgress = 0;
  @observable bannerProgress = 0;

  constructor(guid) {
    this.guid = guid;
    this.feedStore = new ChannelFeedStore(guid);
  }

  @action
  clear() {
    this.channel = {};
    this.rewards = {};
    this.avatarProgress = 0;
    this.bannerProgress = 0;
    this.isUploading = false;
  }

  @action
  markInactive() {
    this.active = false;
  }

  @action
  setChannel(channel, loaded = false) {
    channel = UserModel.checkOrCreate(channel);
    this.channel = channel;
    this.feedStore.setChannel(channel);
    this.active = true;
    this.guid = channel.guid;

    if (loaded) {
      this.loaded = loaded;
    }
  }

  @action
  async load(defaultChannel) {
    if (defaultChannel) {
      defaultChannel = cloneDeep(defaultChannel);
    }
    const channel = await channelsService.get(this.guid, defaultChannel);

    if (channel) {
      this.loaded = true;
      this.setChannel(channel);
      return channel;
    }
    return false;
  }

  @action
  loadrewards(guid) {
    wireService
      .rewards(guid)
      .then(
        action((rewards) => {
          if (rewards) {
            // merge rewards
            rewards.merged = rewards.money.concat(rewards.points);
            this.rewards = rewards;
          } else {
            this.rewards = {};
          }
        }),
      )
      .catch((err) => {
        logService.exception('[ChannelStore] loadrewards', err);
      });
  }

  @action
  setBannerProgress(value) {
    this.bannerProgress = value;
  }

  @action
  setAvatarProgress(value) {
    this.avatarProgress = value;
  }

  @action
  setIsUploading(value) {
    this.isUploading = value;
  }

  @action
  async uploadAvatar(avatar) {
    this.setIsUploading(true);
    this.avatarProgress = 0;

    let avatarResult;

    try {
      avatarResult = await channelService.upload(
        this.guid,
        'avatar',
        {
          uri: avatar.uri,
          type: avatar.type,
          name: avatar.fileName || 'avatar.jpg',
        },
        (e) => {
          this.setAvatarProgress(e.loaded / e.total);
        },
      );

      this.setAvatarProgress(100);
    } catch (error) {
      this.setAvatarProgress(0);
      throw error;
    } finally {
      this.setIsUploading(false);
    }
  }

  @action
  async save({ avatar, banner, ...data }) {
    this.bannerProgress = 0;

    try {
      if (avatar) {
        await this.uploadAvatar(avatar);
      }

      if (banner) {
        this.isUploading = true;
        const bannerResult = await channelService.upload(
          this.guid,
          'banner',
          {
            uri: banner.uri,
            type: banner.type,
            name: banner.fileName || 'banner.jpg',
          },
          (e) => {
            this.setBannerProgress(e.loaded / e.total);
          },
        );

        this.setBannerProgress(100);

        if (bannerResult.error) return bannerResult.error;
      }

      const result = await channelService.save(data);
      const success = result && result.status === 'success';

      if (success) {
        this.channel.name = data.name;
        this.channel.briefdescription = data.briefdescription;
      }

      return success;
    } catch (e) {
      throw e;
    } finally {
      this.setIsUploading(false);
    }
  }

  @action
  reset() {
    this.guid = null;
    this.feedStore = null;
    // only observable by reference because UserModel already have the needed observables
    this.channel = {};
    this.rewards = {};
    this.loaded = false;
    this.active = false;
    this.isUploading = false;
  }
}
