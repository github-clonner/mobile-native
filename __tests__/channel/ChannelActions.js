import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import ChannelActions from '../../src/channel/ChannelActions';
import UserModel from '../../src/channel/UserModel';
import userFaker from '../../__mocks__/fake/channel/UserFactory';
import ChannelStore from '../../src/channel/ChannelStore';
import features from '../../src/common/services/features.service';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { getStores } from '../../AppStores';

jest.mock('../../src/channel/ChannelStore');
jest.mock('../../src/common/services/features.service');
jest.mock('../../AppStores');
jest.mock('../../src/common/services/boosted-content.service');

getStores.mockReturnValue({
  user: {
    me: {},
    load: jest.fn(),
    setUser: jest.fn(),
  },
});

/**
 * Tests
 */
describe('channel actions component', () => {
  let store;

  beforeEach(() => {
    store = new ChannelStore();
    store.channel = new UserModel(userFaker(1));
    store.channel.toggleSubscription = jest.fn();
    store.channel.toggleBlock = jest.fn();
  });

  it('should renders correctly', () => {
    // return true to has crypto
    features.has.mockReturnValue(true);

    const component = renderer
      .create(<ChannelActions store={store} />)
      .toJSON();

    expect(component).toMatchSnapshot();
  });

  it('should show the correct options', () => {
    const wrapper = shallow(<ChannelActions store={store} />);

    let opt = wrapper.instance().getOptions();

    expect(opt).toEqual(['Cancel', 'Block', 'Report']);

    // if subscribed
    store.channel.subscribed = true;
    opt = wrapper.instance().getOptions();

    expect(opt).toEqual(['Cancel', 'Unsubscribe', 'Block', 'Report']);

    // if blocked
    store.channel.blocked = true;
    opt = wrapper.instance().getOptions();
    expect(opt).toEqual(['Cancel', 'Unsubscribe', 'Un-Block', 'Report']);
  });

  it('should show run the correct option', () => {
    const navigation = { push: jest.fn() };
    const wrapper = shallow(
      <ChannelActions store={store} navigation={navigation} />,
    );

    store.channel.subscribed = true;
    store.channel.blocked = true;

    opt = wrapper.instance().executeAction(1);
    expect(store.channel.toggleSubscription).toBeCalled();

    opt = wrapper.instance().executeAction(2);
    expect(store.channel.toggleBlock).toBeCalled();

    opt = wrapper.instance().executeAction(3);
    expect(navigation.push).toBeCalled();
  });
});
