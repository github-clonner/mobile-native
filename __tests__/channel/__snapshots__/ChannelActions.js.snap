// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`channel actions component should renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
        "justifyContent": "flex-end",
      },
      Object {
        "marginTop": 10,
      },
    ]
  }
>
  <View
    accessibilityLabel="More"
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#5DBAC0",
        "borderRadius": 36,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginLeft": 5,
        "opacity": 1,
      }
    }
    testID="MoreButton"
  >
    <Text
      style={
        Array [
          Object {
            "color": "white",
          },
          Array [
            Object {
              "marginLeft": 5,
            },
            Object {
              "marginRight": 5,
            },
          ],
        ]
      }
    >
       
      More
       
    </Text>
  </View>
</View>
`;
